File: pandas/general_query_19.py
#!/usr/bin/python3

"""General query 19

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València

Distribution of exchange students in terms of their occupation
(study or internship/placement) and level (graduate or postgraduate).
Therefore, four groups are considered: graduate students, graduate interns,
postgraduate students and postgraduate interns.

pandas - general_query_19.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student
from util.functions import round_half_down

from sys import argv

def mobility_type_study_level_distribution(students_df):
    students_df = students_df.dropna(subset=['MOBILITYTYPE', 'LEVELSTUDY'])
    mtls_groups_df = students_df.groupby(['MOBILITYTYPE', 'LEVELSTUDY'], as_index=False, sort=False).size()
    mtls_groups_df['MOBILITYTYPE'] = mtls_groups_df['MOBILITYTYPE'].transform(lambda mt: mt.value)
    mtls_groups_df['STUDYLEVEL'] = mtls_groups_df['LEVELSTUDY'].apply(lambda sl: sl.value)

    num_students = mtls_groups_df['size'].sum()
    mtls_groups_df['proportion'] = mtls_groups_df['size'].apply(lambda mtls: round_half_down(mtls / num_students))

    mtls_groups_df = mtls_groups_df[['MOBILITYTYPE', 'STUDYLEVEL', 'proportion']]
    mtls_groups_df = mtls_groups_df.set_index(['MOBILITYTYPE', 'STUDYLEVEL']).sort_index().squeeze()
    mtls_groups_df.index = mtls_groups_df.index.map(lambda mtls: '{}\t{}'.format(*mtls))
    return mtls_groups_df

# python3 general_query_19.py ../data/Student_Mobility.csv [out/pandas19-output]
if __name__ == "__main__":
    compute_task(mobility_type_study_level_distribution, Student, argv[1:])

File: spark/src/main/java/systematicchaos/studentmobility/GeneralQuery19.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * General query 19: Distribution of exchange students in terms of their occupation
 * (study or internship/placement) and level (graduate or postgraduate).
 * Therefore, four groups are considered: graduate students, graduate interns,
 * postgraduate students and postgraduate interns.
 * 
 * spark - GeneralQuery19.java
 */

package systematicchaos.studentmobility;

import static org.apache.spark.sql.functions.callUDF;
import static org.apache.spark.sql.functions.sum;
import static systematicchaos.studentmobility.util.Functions.roundHalfDown;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.api.java.UDF1;
import org.apache.spark.sql.types.DataTypes;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.spark.SparkJob;
import systematicchaos.studentmobility.spark.SparkJobLauncher;
import systematicchaos.studentmobility.spark.SparkTask;
import systematicchaos.studentmobility.util.KeyCountPair;
import systematicchaos.studentmobility.util.StudentBuilder;

public class GeneralQuery19 {
	
	public static class MobilityTypeStudyLevelDistribution implements SparkTask<Double> {
		
		@Override
		public Iterable<KeyCountPair<Double>> computeSparkTask(String[] inputFiles) {
			SparkJob<Student> job = new SparkJob<>(GeneralQuery19.class.getName(),
					Student.class, new StudentBuilder());
			List<KeyCountPair<Double>> mobilityTypeStudyLevel = null;
			
			try {
				Dataset<Student> students = job.readDataset(inputFiles);
				students = students.filter(new Column("mobilityType").isNotNull())
						.filter(new Column("studyLevel").isNotNull());
				
				Dataset<Row> mltsGroups = students.groupBy("mobilityType", "studyLevel").count()
						.orderBy("mobilityType", "studyLevel");
				long numStudents = mltsGroups.agg(sum("count")).first().getLong(0);
				
				mltsGroups.sqlContext().udf().register("MLTS_PROPORTION",
						(UDF1<Long, Double>)(count) -> {
							return (double) count / numStudents;
						}, DataTypes.DoubleType);
				mltsGroups = mltsGroups.withColumn("proportion", callUDF("MLTS_PROPORTION", new Column("count")));
				
				mobilityTypeStudyLevel = mltsGroups.takeAsList((int) mltsGroups.count()).stream()
						.map((mlts) -> new KeyCountPair<>(
								String.format("%s\t%s", Student.MobilityType.valueOf(mlts.getString(0)).getValue(),
										Student.StudyLevel.valueOf(mlts.getString(1)).getValue()),
								roundHalfDown(mlts.getDouble(3))))
						.collect(Collectors.toList());
			} catch (Exception e) {
				System.err.println(e.getClass().getName()+ ": " + e.getMessage());
			} finally {
				job.close();
			}
			
			return mobilityTypeStudyLevel;
		}
	}
	
	// spark-submit --class systematicchaos.studentmobility.GeneralQuery19 --master local[*] target/spark-1.0.0.jar ../data/Student_Mobility.csv [out/spark19-output]
	public static void main(String[] args) throws Exception {
		SparkJobLauncher.launchSparkJob(new MobilityTypeStudyLevelDistribution(), args);
	}
}
